### ISSUE TITLE
[Question] Does asyncssh support pkcs #11 based authentication?

### ISSUE THREAD

**Issue by krishna-veerareddy** [ronf/asyncssh#298]
Hey @ronf,

Does asyncssh support smartcard based authentication similar to this [feature request in paramiko](https://github.com/paramiko/paramiko/pull/827/) where a PKCS11 session can be passed in to the connect call? I can currently work around this by adding the keys to the agent using the `add_smartcard_keys` method but this requires an SSH agent to be running.

I tried using the `load_resident_keys` method that was recently added but my smartcard(Yubikey 4) doesn't support FIDO2 and I run into an "CTAP error: 0x01 - INVALID_COMMAND". Any ideas on how I can get around adding keys to the SSH agent?

Also, thank you for all your work on this project! It is amazing!!

Regards,
Krishna Veera Reddy

**Comment by ronf** [comment]
Right now, as you discovered, AsyncSSH only supports directly accessing smart cards which are capable of supporting U2F or FIDO2. PKCS#11 smart cards operating in PIV mode are also supported, but only indirectly through an ssh-agent. You can authenticate with PKCS#11 PIV keys in AsyncSSH and even add or remove smart card providers in the agent using AsyncSSH, but you have to be running an ssh-agent process.

I haven't really looked into what it would take to call into the PKCS#11 provider library directly from AsyncSSH, rather than relying on the agent to do this. I'd need to read up on the PKCS#11 APIs to judge how much code that would be.

Is there a reason you don't want to use ssh-agent here? One of the benefits it provides is that you don't have to enter the PIN every time you want to use a key. If I were to add support to call into the PKCS#11 library directly, I think it'd probably require that a PIN is passed in explicitly on each connect() call.

**Comment by krishna-veerareddy** [comment]
Hey Ron, thank you for responding.

The reason I didn't want to use an SSH agent is I have a long running script and the agent adds another dependency that could go down or keys could be removed from it by the user causing the script to fail. Also, some of my users don't have an agent running so I need to account for that as well.

So, instead of worrying about these cases I thought I'd explore ways to not have to use an SSH agent but if it's not something asyncssh currently provides then that's totally fine. Just wanted to make sure I am not missing something obvious.

**Comment by ronf** [comment]
Makes sense. To be honest, I wasn't sure PKCS#11 was something used very often with SSH, so once I had it working with ssh-agent I didn't really think much about implementing direct support for it. The direct U2F/FIDO2 support came in much more recently, but I never revisited the PKCS#11 support to consider doing that.

I did some preliminary exploration of this today, and it looks like there's a pretty nice pkcs11 module for Python that does most of the heavy-lifting. From an interactive shell, I've had no trouble pulling public keys off of a PIV card and converting them into a form AsyncSSH can use. I've also confirmed that I can sign a block of data using the private key on the card and then verify it with my existing AsyncSSH code. So, it might not be too difficult to add this, with the Python pkcs11 module being an optional decency for people who want that.

**Comment by krishna-veerareddy** [comment]
Yeah makes sense to have pkcs11 as an optional dependency since the usage of this feature seems uncommon. Hey if you think this is a worthy addition to asyncssh and you aren't able to get to it anytime soon let me know and I can take a look at this as well. I'd be happy to help in anyway I can.

**Comment by ronf** [comment]
Thanks very much - I could definitely use your help in doing some testing when I've got something to try. I had a little time this week to experiment, and I have working code that's accessing a PIV card and loading key handles and public keys from it, as well as code that's using those keys to sign data that I can verify successfully using the corresponding public key. So, now I just need to wrap that up in child class of SSHKeyPair, similar to the SSHAgentKeyPair.

I'm also still thinking about how to expose this in the AsyncSSH API. I see there's an OpenSSH config file variable called PKCS11Provider, so I'll probably enable support for that, along with a corresponding SSHClientConnectionOptions argument. However, I'll also want to provide some way to specify which token(s) to use and to provide PIN information for those tokens. In the simplest (and most common) case, there'd be a single token, so I'll probably just have something like a pkcs11_pin argument that can be a simple string. However, I'm thinking I might also allow that to be something like a dictionary where the keys could be token labels or serial numbers and the values could be the PIN to use for each of those tokens. That way, if you wanted to load keys from specific tokens you could, and you could even load keys from multiple tokens with different PINs. It might also be useful to be able to limit which keys to load from a token, though, based on either key ID or label.

Perhaps the more advanced filtering should be done with an explicit call to some kind of load_pkcs11_keys() function. Once you get the keys you want, could pass them into connect() via the existing client_keys argument. However, for the simple case AsyncSSH could load available keys automatically if you specify pkcs11_provider and pkcs11_pin arguments to connect(), much like how it works with the agent_path argument today.

**Comment by ronf** [comment]
A quick update on this: I have working code that can retrieve key information off of a PKCS#11 token and use those keys as client keys to perform public key authentication in the client, and I tested this against OpenSSH.

Today, I actually extended that code to also load X.509 certificates off of the token as well, so you can optionally use X.509 certificates instead of plain private/public key pairs to authenticate against SSH servers that support that (e.g. pkixssh). This also works against AsyncSSH when it is used as a server.

Unfortunately, I've run into a bit of a snag with the PKCS11 library where it blocks the entire process when performing operations, including waiting for the user to touch their security token when that's enabled. I tried putting this code in an executor to work around this (like i do for FIDO auth), but unfortunately the library call ends up blocking all threads. I'm guessing it may be holding onto the Python global interpreter lock when it blocks, and I'm not sure that's something I'll be able to easily fix.

One option here would be to fork off a separate process within AsyncSSH for performing these operations, but that's a messy thing to try and do from a library when you have no idea what resources might be opened already by the application before AsyncSSH gets a chance to run. So, I don't know how viable that would be.

There's other work I could do around being more selective about what keys to load and allowing this things like server host keys to be stored on a security token. However, until I can come up with a plan for dealing with this blocking issue, I'm hesitant to check anything in.

**Comment by ronf** [comment]
Good news - I think I may have figured out how to prevent the python-pkcs11 code from blocking all threads while doing a signing operation, allowing the executor in AsyncSSH to work properly. See danni/python-pkcs11#86 for more details on what I have come up with. This change may be needed in other places as well, but for AsyncSSH's specific use of PCKS#11, I think signing may be the only place where it matters right now.

**Comment by krishna-veerareddy** [comment]
This is amazing! I really appreciate all the effort you are putting into this!!

**Comment by ronf** [comment]
Another update: Tonight, my pull request at danni/python-pkcs11#87 was accepted (thanks, Danni!), addressing the blocking problem I was seeing in the python-pkcs11 package. The new PKCS#11 client code in AsyncSSH is also pretty much ready to go. I'm just working on some stubs to allow me to add unit tests for it, and then it should be ready to check in, hopefully by this weekend.

The API changes are pretty much as I described above, except that I backed off from allowing the PIN to be specified as a dictionary to accommodate loading keys from multiple tokens with different PINs. That was getting too messy. Instead, I provided a load_pkcs11_keys() function that lets you be selective about which tokens to load keys from and which keys to load. You can call that multiple times to collect the keys you want, passing those into AsyncSSH connect() via the client_keys argument. In the simple case, though, you just need to provide a pkcs11_provider and a pkcs11_pin in the connect arguments and AsyncSSH will automatically load all the keys it finds.

**Comment by ronf** [comment]
It took a bit longer than expected to complete the unit tests for this feature, but an initial cut of PKCS#11 support is now available in the "develop" branch as commit b570941. As it turned out, the unit test code was actually bigger than the feature itself!

To use this feature, you can either specify pkcs11_provider and pkcs11_pin arguments in the connect() call, or you can use the new load_pkcs11_keys() function, which allows you to load only from specific tokens or load only specific keys. It also lets you specify different pins for different tokens (by calling the function multiple times). Once you have all the keys, you can pass them into connect() via the client_keys argument.

Another feature in this commit is the ability to authenticate using X.509 certificates stored on the security token. OpenSSH doesn't support X.509 certificates, but this can work when using AsyncSSH on both ends, or with other SSH server implementations such as PKIX-SSH.

For now, I've only added support for RSA and ECDSA keys, as that's all my Yubikey supports. If you need other key types, let me know. I'm hoping to eventually support Ed25519 keys, but I've only seen support for those so far in much more expensive HSM devices.

**Comment by ronf** [comment]
I just found and fixed a minor bug in the last checkin, and also updated setup.py to request version 0.7.0 of python-pkcs11 as an optional dependency. This version includes the fix I made to not hold the global interpreter lock when making blocking calls to the PKCS#11 provider functions.

**Comment by krishna-veerareddy** [comment]
Sounds good! Will check this out tonight.

**Comment by krishna-veerareddy** [comment]
@ronf Hey I tested it out on a sample program and it works flawlessly! I tested out both the `load_pkcs11_keys` option and the passing the provider/pin to the connect call. I haven't tested the X.509 certificates based authentication yet although I don't think it will work in my case but will give it a try.

**Comment by krishna-veerareddy** [comment]
Hey what are your thoughts on removing the option of passing `pkcs11_provider` and `pkcs11_pin` options directly to the connect call? For cases where we have to jump connect through another host we would have to pass the provider and pin to both the connect calls:

```python
async with asyncssh.connect(bastion, pkcs11_provider = '...', pkcs11_pin = '...') as tunnel:
        async with tunnel.connect_ssh(host, pkcs11_provider = '...', pkcs11_pin = '...') as connection:
            result = await connection.run('whoami')
```

I feel the load_pkcs11_keys option is good enough but would love to know your thoughts on it.

**Comment by ronf** [comment]
Glad to hear it is working for you! :)

If you choose to use load_pkcs11_keys()l you actually don't need to pass the pkcs11_provider and pkcs11_pin arguments to connect() to be able to use the keys that you explicitly loaded. So, you should be good to go on both of the calls in the jump host case if you pass the keys via the client_keys argument in both calls, without pkcs11_provider/pkcs11_pin.

You can also call load_pkcs11_keys() multiple times with different token_label or token_serial values and PINs and combine the resulting set of keys together in later calls to connect().

**Comment by krishna-veerareddy** [comment]
> If you choose to use load_pkcs11_keys(), you actually don't need to pass the pkcs11_provider and pkcs11_pin arguments to connect() to be able to use the keys that you explicitly loaded. So, you should be good to go on both of the calls in the jump host case if you pass the keys via the client_keys argument in both calls, without pkcs11_provider/pkcs11_pin.

I understand that but wanted to mention that the option of passing `pkcs11_provider` and `pkcs11_pin` maybe inferior to the `load_pkcs11_keys` since with the former method would need to load the keys(blocking?) from the smartcard for each connect call, right? I could have totally got that wrong though as I sift through the changes pretty quickly. In any case `load_pkcs_keys` works for me. Thank you for all the work you put into this!!

**Comment by ronf** [comment]
In the case where you want to make multiple connections, you're right that it would be more efficient to load the keys explicitly ahead of time. That also applies to all of the other cases where AsyncSSH will load configuration for you, though, such as reading keys out of the default "~/.ssh" pathnames, or reading them from SSH agent. Even passing in an explicit filename in something like "client_keys" is less efficient than doing an explicit load_keypairs(), if you are going to use the keys more than once.

The option to have AsyncSSH load things for you is intended for the simple case where you are making a single client connection, or where you are setting up a single server, or where you want to pick up the latest values every time. That's still a pretty common case, and the goal with pkcs11_provider/pkcs11_pin is to make it easy for that to all happen in a single call, just like the other key-related connect() arguments.

If you are looking to optimize for multiple connections, you might want to also think about explicitly reading the known_hosts file in once, assuming you're ok with not picking up any updates made to it for the lifetime of your process (or until you explicitly reload things). That's the flip-side of the efficiency argument -- loading things yourself means you are locking in that set of values, which is both good and bad.

Regarding your question about blocking, it's true that the load operation is "blocking", though not for very long. If you don't want that, you can explicitly run the load_pkcs11_keys() call (or any other blocking load/read calls) in an executor. AsyncSSH does this automatically for the sign() operation for both PKCS#11 and U2F/FIDO keys, as that can block waiting for the user to touch the security token. By using an executor, other asyncio operations can proceed while waiting for the sign operation to complete and only the particular connect() call waiting for you to touch the key is blocked.

**Comment by ronf** [comment]
This functionality is now released in AsyncSSH 2.4.0.
